bpo-41100: Support macOS 11 and Apple Silicon

ronaldoussoren opened the pull request:
This merges my other two PRs, and improves on the PR that implements support for older macOS versions.

- [x] Tests pass on macOS 11 (ARM64)
- [x] Tests pass on macOS 10.9 using 10.10 SDK
- [x] Add tests for the time module, similar to the tests added for the posix module
- Build arm64/x86_64 binary ("universal2") and test on:
   - [x] macOS 11 (arm64)
   - [ ] macOS 11 (x86_64)
   - [x] macOS 10.15 (x86_64)
   - [x] macOS 10.14 (x86_64)
   - [ ] macOS 10.12 (x86_64)
   - [x] macOS 10.9 (x86_64)
- [x] _osx_support.py changes to strip "-arch arm64" (and possibly related arguments)
       on systems that don't support his.
- [x] Make sure CI passes on all platforms
- [ ] Add tests for the changes in _osx_support and distutils
- [ ] Write NEWS blurb

<!-- issue-number: [bpo-41100](https://bugs.python.org/issue41100) -->
https://bugs.python.org/issue41100
<!-- /issue-number -->

ronaldoussoren committed:
Add support for ARM64 to the pythonw tool

ronaldoussoren committed:
Add support for "universal2" as a fat binary target on macOS

ronaldoussoren committed:
Tweak ctypes.macholib.dyld to work with the shared library cache.

ronaldoussoren committed:
Silence compile time warning

We should consider just dropping support for macOS 10.4 here
to simplify the code.

ronaldoussoren committed:
macOS/arm64 support, based on GH-21249

This is support for ctypes on macOS/arm64 based
on PR 21249 by Lawrence D'Anna (Apple).

Changes:
- changed __builtin_available tests from 11.0 to 10.15
- added test to setup.py for ffi_closure_alloc and use
  that in malloc_closure.c
- Minor change in the code path for ffi_prep_closure_var
  (coding style change)

ronaldoussoren committed:
Fix test for macOS before 10.10

ronaldoussoren committed:
ARM64 is a valid platform for macOS

ronaldoussoren committed:
Unconditionally use uint32_t here.

The preprocessor guard in the old version doesn't work, and
isn't really needed (10.4 only supported 32-bit code where
unsigned long is the same as uint32_t).

ronaldoussoren committed:
First stab at deploying to an earlier version of macOS than the build machine

ronaldoussoren committed:
Update build-installer.py for universal2 builds

This also adds an option to stop building after compiling the
3th-party dependencies, as well as a script for archiving those
dependencies. This makes it easier to work on the build.

There are three changes to build-installer.py related to universal2
support:

1. Add 'universal2' information to build-installer.py;
2. Building OpenSSL for arm64 requires a patch at this time;
3. For some reason I had to patch the Tcl build to avoid a build error.

ronaldoussoren committed:
Merge ctypes changes from Apple (PR 21241)

Needed because my previous workaround doesn't work anymore.

This uses a private API, that should be made public later...

ronaldoussoren committed:
The archive-deps.py script isn't really useful

ronaldoussoren committed:
Merge branch 'macos-deploy-to-earlier' into macos11-deploy-earlier-branch

ronaldoussoren committed:
Cleanup the code to check at runtime for a couple of APIs.

ronaldoussoren committed:
Dynamicly fill posix._have_functions as well

ronaldoussoren committed:
Changes after testing with Xcode 6.2 on macOS 10.9

- __builtin_available is not present in the compiler, fall
  back to the ancient way of testing if a symbol is usable.
- Logic error code in pytime.c resulted in a compile error.

ronaldoussoren committed:
Code formatting

ronaldoussoren committed:
Fix a number of errors in weaklinking support

ronaldoussoren committed:
Merge branch 'macos11-deploy-earlier-branch' of github.com:ronaldoussoren/cpython into macos11-deploy-earlier-branch

ronaldoussoren committed:
Implement tests for weak linked symbols.

The tests are not entirely done yet, and something
similar needs to be done for the time module.

ronaldoussoren committed:
Finish the test for os.link behaviour

ronaldoussoren committed:
posixmodule.c uses multi-phase initialization

ronaldoussoren committed:
Make sure some form of prep_cif is called, even if the variadic version isn't present.

I don't particularly like the code duplication, but this code
should be pretty stable.

ronaldoussoren committed:
Move the removal of unavailable functions to the module exec

ronaldoussoren committed:
Improved error handling

ronaldoussoren committed:
Remove '-arch arm64' from the compiler arguments when necessary

This patch is not 100% correct: It ignores Xcode 12.2 on macOS 10.15.

maxbelanger commented on Modules/_ctypes/callproc.c:
```diff
@@ -57,13 +57,27 @@
+ * - Switch to that API
+ * - Add feature macro and runtime guards  (as with the ffi.*loc API's)
+ */
+extern bool _dyld_shared_cache_contains_path(const char* path) __attribute__((weak_import));
```
This is now defined publicly in the SDK w/ the proper availability declaration; including `mach-o/dyld.h` should work.

ronaldoussoren replied:
Indeed, I'll remove the local declaration.

maxbelanger commented on Modules/timemodule.c:
```diff
@@ -1373,6 +1398,16 @@ _PyTime_GetThreadTimeWithInfo(_PyTime_t *tp, _Py_clock_info_t *info)
 #elif defined(HAVE_CLOCK_GETTIME) && defined(CLOCK_PROCESS_CPUTIME_ID)
 #define HAVE_THREAD_TIME
+
+#ifdef __APPLE__
```
Do we also want to check for `__has_attribute(availability)` here?

ronaldoussoren replied:
I guess so. The API is new in 10.13 and the version of Xcode with the 10.13 SDK always supports this attribute, but someone might compile with GCC.

maxbelanger commented on setup.py:
```diff
@@ -2149,11 +2138,10 @@ def detect_ctypes(self):
             sources.append('_ctypes/malloc_closure.c')
-            sources.append('_ctypes/darwin/dlfcn_simple.c')
+            extra_compile_args.append('-DUSING_MALLOC_CLOSURE_DOT_C=1')
+            #sources.append('_ctypes/darwin/dlfcn_simple.c')
```
I don't suppose that's needed anymore?

ronaldoussoren replied:
Correct.

maxbelanger commented on setup.py:
```diff
@@ -2100,43 +2114,18 @@ def detect_tkinter(self):
-        self.use_system_libffi = False
+
+        if (not sysconfig.get_config_var("LIBFFI_INCLUDEDIR") and MACOS and
+            (is_macosx_at_least((10,15)) or '-arch arm64' in sysconfig.get_config_var("CFLAGS"))):
```
I'm not sure we should be choosing to use the system `libffi` based on the deployment target, since the code changes you've made make it resilient to earlier targets via availability checks. Why not make it based on the SDK version?

ronaldoussoren replied:
I'm removing the second line of the test, the FFI headers are available on 10.9 (which is as far back as I can test).

maxbelanger commented on Python/pytime.c:
```diff
@@ -683,15 +683,22 @@ pygettimeofday(_PyTime_t *tp, _Py_clock_info_t *info, int raise)
 
 #ifdef HAVE_CLOCK_GETTIME
+
+#if defined(__APPLE__) 
```
Do we want to check for `__has_builtin(__builtin_available)` here?

maxbelanger reviewed (commented):
Are we still planning on backporting this to change to Python 3.8?

ronaldoussoren committed:
check that weak linking works for the time module

ronaldoussoren committed:
Drop the DEPS_ONLY changes

This change is independent of the macOS 11/arm64 changes.

ronaldoussoren commented on Mac/BuildScript/build-installer.py:
```diff
@@ -190,6 +196,27 @@ def getTargetCompilers():
     return getDeptargetTuple() >= (10, 6)
 
+
+def tweak_tcl_build(basedir, archList):
```
@ned-deily  I'm not too happy about this, for some reason I needed this to get a working build.

ned-deily replied:
@ronaldoussoren build-installer's Tcl/Tk build steps need to be tweaked for building on 10.15+. 8.6.8 built on 10.9 executes reasonably well on newer systems but 8.6.8 build on, say, 10.15 doesn't.  I'll hack on that after we get the rest in.

ronaldoussoren committed:
Undo change to default PATH

ronaldoussoren commented on Mac/BuildScript/openssl-mac-arm64.patch:
```diff
@@ -0,0 +1,41 @@
+diff -ur openssl-1.1.1g-orig/Configurations/10-main.conf openssl-1.1.1g/Configurations/10-main.conf
```
I've checked the OpenSSL repo and it does not yet contain a similar patch in the 1.1.1 branch. This patch is still needed.

ronaldoussoren commented on Modules/_ctypes/callbacks.c:
```diff
@@ -408,13 +409,35 @@ CThunkObject *_ctypes_alloc_callback(PyObject *callable,
+#   if USING_APPLE_OS_LIBFFI
+#      define HAVE_FFI_PREP_CLOSURE_LOC_RUNTIME __builtin_available(macos 10.15, ios 13, watchos 6, tvos 13, *)
+#   else
+#      define HAVE_FFI_PREP_CLOSURE_LOC_RUNTIME true
```
Maybe use "1" instead of "true"?

serhiy-storchaka replied:
You need `#include <stdbool.h>` for `true`.

ronaldoussoren commented on Modules/posixmodule.c:
```diff
@@ -7,18 +7,6 @@
    of the compiler used.  Different compilers define their own feature
    test macro, e.g. '_MSC_VER'. */
 
-#ifdef __APPLE__
```
The side effect of this is to drop support for building on 10.5 and deploying to 10.4. That's such an esoteric use case that I'm in favour of doing this even in back ports to earlier versions.

ronaldoussoren commented on Modules/posixmodule.c:
```diff
@@ -51,6 +39,125 @@
+ *    if (HAVE_<FUNCTION>_RUNTIME) { ... }
+ *
+ * In mixing the test with other tests or using negations will result in compile 
+ * errors.
```
An assumption of the HAVE_<FUNCTION>_RUNTIME available is that using "if (1) {... }" is zero-cost, IMHO that's a reasonable assumption.

ronaldoussoren committed:
Use configure to detect the presence of _dyld_shared_cache_contains_path

ronaldoussoren committed:
Revert unnecessary change

ronaldoussoren committed:
Updates to timemodule.c

- Check availability of __attribute__((availability
- Revert unneccessary change to module init function

ronaldoussoren committed:
Stricter checking of __builtin_available

ronaldoussoren committed:
Clean up setup.py

squeaky-pl commented on Lib/_osx_support.py:
```diff
@@ -140,6 +140,20 @@ def _supports_universal_builds():
     return bool(osx_version >= (10, 4)) if osx_version else False
 
+def _supports_arm64_builds():
+    """Returns jTrue if arm64 buids are supported on this system"""
```
```suggestion
    """Returns True if arm64 builds are supported on this system"""
```

ronaldoussoren committed:
Fix build failures on Linux

ronaldoussoren committed:
Fix typo

Co-authored-by: Paweł Piotr Przeradowski <showerproof86@gmail.com>

ronaldoussoren committed:
Remove reference to non-existing probe function

lawrence-danna-apple commented on Modules/_ctypes/callproc.c:
```diff
@@ -57,13 +57,27 @@
+ * private for now, this should change in beta 4.
+ *
+ * TODO:
+ * - Switch to that API
```
this api has been made public.   see updates to https://github.com/python/cpython/pull/21241

ronaldoussoren replied:
That's been changed in the current version of the PR.

lawrence-danna-apple commented on setup.py:
```diff
@@ -2183,31 +2171,48 @@ def detect_ctypes(self):
-            ffi_inc = find_file('ffi.h', [], ffi_inc_dirs)
-        if ffi_inc is not None:
-            ffi_h = ffi_inc[0] + '/ffi.h'
+            # XXX: The define should only be added when actually using the system
```
the comment is correct.  we don't want `__builltin_available` checks for libffi functionality unless the libffi being used is the one bundled with the OS

ronaldoussoren replied:
I agree. There's another bug as well: the code will overwrite ffi_inc if it was set earlier (user has pkg-config installed and a non-system version of libffi, for example in homebrew).  The define should only be set when we're using ffi_in_sdk and that directory exists. 

I'll push an updated version shortly.

ronaldoussoren committed:
One more attempt at building on Linux...

ronaldoussoren committed:
A bit of clean up of the setup.py logic.

- ffi_inc was unconditionally set to the SDK path (even if
  LIBFFI_INCLUDEDIR was set)
- USING_APPLE_OS_LIBFFI should only be set when actually using
  the SDK version

ronaldoussoren committed:
Fix patchcheck issues

ronaldoussoren committed:
add prefix to f-strings

ronaldoussoren committed:
Add missing preprocessor guards

ronaldoussoren committed:
Use the correct cut-off for the availability of clock_gettime and
friends

ronaldoussoren committed:
Fix error in preprocessor logic for os.chmod implementation

ronaldoussoren committed:
Merge branch 'master' into macos11-deploy-earlier-branch

dimaqq commented on Lib/_osx_support.py:
```diff
@@ -140,6 +140,20 @@ def _supports_universal_builds():
 
+def _supports_arm64_builds():
+    """Returns True if arm64 builds are supported on this system"""
+    # There are two sets of systems supporting macOS/arm64 builds:
+    # 1. macOS 11 and later, unconditionally
+    # 2. macOS 10.15 with Xcode 12.2 or later
+    # For now the second category is ignored.
```
I've discovered in another issue that official releases are build on old-ish macos.
Is this patch only for self-built python then?
Or is the intention to release arm and x86 packages separately?
Or perhaps to release "classic x86" and "arm+x86" packages?

ned-deily replied:
We're still working on this so we're not quite ready yet to say exactly what will be rolled out with which releases but our intent is to at least provide a universal installer variant that supports macOS 11.0 Big Sur and earlier systems and both Intel-64 and Apple Silicon natively.

ronaldoussoren replied:
I don't have a definitive answer for your question, but my goal is to make it possible to build the official installers on a recent system.  That has several advantages: (1) it is required for arm64, (2) you can use new APIs when using a new system, (3) newer compilers generate better code.

I'd personally like to see 1 installer for Python 3.10: x86_64+arm64 for macOS 10.9 or later.  I don't know yet what the best path forward is for 3.9 (and possibly 3.8)

And current installers are indeed build on an old system, before this PR the only way to have an installer that works on macOS 10.9 and later is to build using the 10.9 SDK.

maxbelanger replied:
From our perspective, the addition of runtime availability checks as part of this change is fantastic. We ship our own custom fork of Python in our macOS app, which we build using the standard/recommended approach: use the latest Xcode and macOS SDK + set your deployment target to the minimum version you want to support at runtime.

vstinner commented on Lib/test/test_time.py:
```diff
@@ -1041,6 +1041,36 @@ def test_object_to_timespec(self):
+        import platform
+
+        config_vars = sysconfig.get_config_vars()
+        self.available = { nm for nm in config_vars if nm.startswith("HAVE_") and config_vars[nm] }
```
It looks overkill to use a dict since only HAVE_CLOCK_GETTIME is used.

ronaldoussoren replied:
I agree. This was copied from test_posix.py. I've adjusted the code.

vstinner commented on Lib/test/test_time.py:
```diff
@@ -1041,6 +1041,36 @@ def test_object_to_timespec(self):
 
+@unittest.skipUnless(sys.platform == "darwin", "test weak linking on macOS")
+class TestTimeWeaklinking(unittest.TestCase):
+    # These test cases verify that weak linking support on macOS works
```
Please add a bpo number (where I would expect a longer explanation of what "weak linking" is ;-)).

ronaldoussoren replied:
Would a reference to the Mac readme be acceptable?  I'd still have to write the explanation, but that would keep the explanation in the source tree instead of in an issue that won't be updated once this is merged.

ronaldoussoren replied:
There is now a reference to Mac/README.txt where added some text about weak linking.

vstinner commented on Lib/test/test_posix.py:
```diff
@@ -1925,13 +1925,253 @@ def test_posix_spawnp(self):
+
+            # XXX: Check if this test is correct, the implementation of os.utime
+            #      is fairly complicated
+
```
PEP 8: please remove one empty line.

vstinner commented on Lib/test/test_posix.py:
```diff
@@ -1925,13 +1925,253 @@ def test_posix_spawnp(self):
+                os.utime("path", follow_symlinks=False)
+
+            # XXX: Check if this test is correct, the implementation of os.utime
+            #      is fairly complicated
```
Please fix this XXX or remove it.

vstinner commented on Lib/test/test_posix.py:
```diff
@@ -1925,13 +1925,253 @@ def test_posix_spawnp(self):
+
+            finally:
+                support.rmtree(base_path)
+
```
PEP 8: a single empty line is enough here.

vstinner reviewed (commented):
This PR is quite large and so really hard to review. Can you try to split it into multiple PRs? For example:
* ctypes/ffi
* time
* posix
* Mac build script
* etc.

ronaldoussoren committed:
the sid attribute for posix_spawn is only available on recent versions of macOS

ronaldoussoren committed:
Changes due to review by Victor Stinner

ronaldoussoren committed:
Merge branch 'macos11-deploy-earlier-branch' of github.com:ronaldoussoren/cpython into macos11-deploy-earlier-branch

serhiy-storchaka commented on Lib/_osx_support.py:
```diff
@@ -140,6 +140,20 @@ def _supports_universal_builds():
+        try:
+            osx_version = tuple(int(i) for i in osx_version.split('.'))
+        except ValueError:
+            osx_version = ''
```
This code is repeated several times. Would be nice to move it to separate function.

We could even make it more reliable. I do not know what is the format for non-release version, but this code does not work for "11.2rc1" or "11.2-beta2".

ronaldoussoren replied:
I'll look into this.

The code itself is safe, macOS pre-release versions don't have a suffix (that is, the current macOS 11.0.1 beta has "11.0.1" as its version, the only way to see that this is a pre-release version is to compare the build number to that of the final release when that is available).

serhiy-storchaka commented on Lib/_osx_support.py:
```diff
@@ -331,6 +345,13 @@ def compiler_fixup(compiler_so, cc_args):
+        for idx in range(len(compiler_so)):
+            if compiler_so[idx] == '-arch' and compiler_so[idx+1] == "arm64":
+                del compiler_so[idx:idx+2]
+                break
```
Can there be multiple occurrences of `-arch arm64`?

ronaldoussoren replied:
Only when the user has specified the same architecture multiple times.  The compiler ignores multiple occurrences of the same architecture.  I'll drop the "break".

serhiy-storchaka replied:
I don't think that the user will do this intentionally, but I afraid that in generated command lines it can happen. It happens with some commands in our Makefile.

Dropping `break` is not correct because `del` shifts the content. You can write the code in the similar ways to the loop above (lines 340-346). Or iterate in the opposite direction: ` for idx in range(len(compiler_so))[::-1]`, it may be simpler.

serhiy-storchaka commented on Lib/test/test_posix.py:
```diff
@@ -1925,13 +1925,248 @@ def test_posix_spawnp(self):
+                    os.link(source_path, link_path, follow_symlinks=False)
+
+            finally:
+                support.rmtree(base_path)
```
You can also use `support.temp_dir()`.

serhiy-storchaka commented on Lib/test/test_time.py:
```diff
@@ -1041,6 +1041,37 @@ def test_object_to_timespec(self):
+        clock_names = [
+            "CLOCK_MONOTONIC", "clock_gettime", "clock_gettime_ns", "clock_settime",
+            "clock_settime_ns", "clock_getres"]
+
```
```suggestion
```

serhiy-storchaka commented on Mac/BuildScript/build-installer.py:
```diff
@@ -801,6 +828,7 @@ def build_openssl_arch(archbase, arch):
         arch_opts = {
             "i386": ["darwin-i386-cc"],
             "x86_64": ["darwin64-x86_64-cc", "enable-ec_nistp_64_gcc_128"],
+            "arm64": ["darwin64-arm64-cc" ],
```
```suggestion
            "arm64": ["darwin64-arm64-cc"],
```

serhiy-storchaka commented on Mac/README.rst:
```diff
@@ -120,6 +120,8 @@ support ppc (Xcode 4 on 10.6 and later systems).  The flavor can be specified
 using the configure option ``--with-universal-archs=VALUE``. The following
 values are available:
 
+  * ``universal2``: ``arm64``, ``x86_64``
```
Should not `all` also contain `arm64`?

ronaldoussoren replied:
I don't want to change the meaning of existing sets.   It is also not possible to build fat binaries with arm64 and ppc or i386 with a single compiler, the current compiler in Xcode only supports arm64 and x86_64.

ronaldoussoren replied:
For 3.10 we can reconsider the sets of architectures to support, possibly defaulting to "universal2" and only keeping that and "intel" (for folks still needing 32-bit support).  But that's for another issue.

serhiy-storchaka commented on Modules/_ctypes/callproc.c:
```diff
@@ -1398,6 +1460,39 @@ copy_com_pointer(PyObject *self, PyObject *args)
+         if (PyBytes_Check(name2))
+             name_str = PyBytes_AS_STRING(name2);
+         else
+             name_str = PyByteArray_AS_STRING(name2);
```
`PyUnicode_FSConverter()` returns bytes. It never returns bytearray.

serhiy-storchaka commented on Modules/_ctypes/callproc.c:
```diff
@@ -1398,6 +1460,39 @@ copy_com_pointer(PyObject *self, PyObject *args)
+         if(_dyld_shared_cache_contains_path(name_str))
+             Py_RETURN_TRUE;
+         else
+             Py_RETURN_FALSE;
```
There is a reference leak. Add `Py_DECREF(name2)` before return.

serhiy-storchaka commented on Modules/posixmodule.c:
```diff
@@ -2362,6 +2471,10 @@ posix_do_stat(PyObject *module, const char *function_name, path_t *path,
     STRUCT_STAT st;
     int result;
 
+#if defined(HAVE_FSTATAT)
```
```suggestion
#ifdef HAVE_FSTATAT
```

serhiy-storchaka commented on Modules/posixmodule.c:
```diff
@@ -2844,17 +2973,40 @@ os_access_impl(PyObject *module, path_t *path, int mode, int dir_fd,
         result = access(path->narrow, mode);
     Py_END_ALLOW_THREADS
+
+#if defined(HAVE_FACCESSAT)
```
```suggestion
#ifdef HAVE_FACCESSAT
```

ronaldoussoren committed:
inor changes

- Review comments by Serhiy Storchaka
- Test_posix utime tests were ignored by testing for the wrong
  feature macro. This uncovered a problem with the test case
  when running on macOS 10.9
- Added blurb

ned-deily reviewed (commented):
See comment in the discussion about current testing results.

ddl-jameslee commented on Lib/_osx_support.py:
```diff
@@ -331,6 +355,12 @@ def compiler_fixup(compiler_so, cc_args):
 
+    elif not _supports_arm64_builds():
+        # Look for "-arch arm64" and drop that
+        for idx in range(len(compiler_so)):
```
When `idx` is `len(compiler_so)-1`, it will throw an `IndexError` exception.
So, it should be `for idx in range(len(compiler_so)-1):`

smoofra replied:
dang, looks like you're right

ned-deily replied:
This PR was merged over a year ago and has long since been released to the field. Please open a new issue on the bug tracker and, if possible, provide a PR.  Otherwise, it will likely be forgotten about it. Thanks!
